src/ParseFasm.cpp

```diff
@@ -0,0 +1,508 @@
+        static constexpr int kWidth = sizeof(num) * 8;
+};
+
+/// Output stream operator for Num.
+/// In hex mode, Nums are printed as <XX>, otherwise they are
+/// copied into the output stream in the underlying representation.
+/// As such, this will use the native endianness.
+std::ostream& operator<<(std::ostream& s, const Num& num) {
+        if (num.tag)
+                s << num.tag;
+        if (hex_mode) {
+                s << "<" << std::hex << num.num << ">";
+        } else {
+                s.write(reinterpret_cast<const char*>(&num.num),
+                        sizeof(num.num));
+        }
+        return s;
+}
+
+/// The Str class wraps a std::string to provide an << operator
+/// that includes the tag and length.
+struct Str {
+        /// Takes the tag and string data
+        Str(char tag, std::string data) : tag(tag), data(data) {}
+
+        char tag;
+        std::string data;
+};
+
+/// This output stream operator adds the header needed to decode
+/// a string of unknown length and type.
+/// Note that some characters are escaped in hex mode to
+/// avoid confusion.
+std::ostream& operator<<(std::ostream& s, const Str& str) {
+        s << str.tag << Num(str.data.size());
+        if (hex_mode) {  /// escape < \ >
+                for (char c : str.data) {
+                        if (c == '<' || c == '>' || c == '\\') {
+                                s.put('\\');
+                        }
+                        s.put(c);
+                }
+        } else {
+                s << str.data;
+        }
+        return s;
+}
+
+/// Wraps a string in another header, used to aggregate data.
+std::string withHeader(char tag, std::string data) {
+        std::ostringstream header;
+        header << tag << Num(data.size());
+        return header.str() + data;
+}
+
+/// Counts characters that don't match the given character.
+/// Used to count digits skipping '_'.
+int count_without(std::string::iterator start,
+                  std::string::iterator end,
+                  char c) {
+        int count = 0;
+        auto it = start;
+        while (it != end) {
+                if (*it != c) {
+                        count++;
+                }
+                it++;
+        }
+        return count;
+}
+
+/// Calculates the number of leading pad bits are needed
+/// so that the rightmost bit will be the LSB of a Num.
+/// e.g. This would be 31 for 33'b0.
+int lead_bits(int bits) {
+        return (Num::kWidth - (bits % Num::kWidth)) % Num::kWidth;
+}
+
+// clang-format off
+/// Decode a hex digit.
+int from_hex_digit(char c) {
+    int result =
+        c >= '0' && c <= '9' ? c - '0' :
+        c >= 'a' && c <= 'f' ? c - 'a' :
+        c >= 'A' && c <= 'F' ? c - 'A' : -1;
+    assert(result >= 0 && result < 16);
+    return result;
+}
+// clang-format on
+
+/// Makes tags easy to extract for documentation and code generation.
+/// Use at most once per line to allow simple grepping.
+#define TAG(c, long_name) (c)
```

Comment:
Does it make sense to do this X-macro style?

E.g.

```
#define TAGS \
  TAG('#', kComment) \
  TAG('f', kFeature)

#define TAG(c, long_name) constexpr char long_name = c;
TAGS
#undef TAGS
```

Reply:
The consumer is Python, so this doesn't apply. See: https://github.com/SymbiFlow/fasm/pull/28/files#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4R44-R54

Nonetheless, even for C code, I prefer including inline over X-macros, because it's cleaner and there are less problems with escaping.